src/glass/src/app/shared/services/api/local-inventory.service.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Project Aquarium's frontend (glass)
+ * Copyright (C) 2021 SUSE, LLC.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+import { Injectable } from '@angular/core';
+import { BehaviorSubject, Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+import {
+  Inventory,
+  LocalNodeService,
+  Nic,
+  NodeStatus,
+  Volume
+} from '~/app/shared/services/api/local.service';
+import { PollService } from '~/app/shared/services/poll.service';
+
+export type NICEntry = {
+  name: string;
+  info: Nic;
+};
+
+@Injectable({
+  providedIn: 'root'
+})
+export class LocalInventoryService {
+  public inventory$: Observable<Inventory | undefined>;
+
+  private inventorySource: BehaviorSubject<Inventory | undefined> = new BehaviorSubject<
+    Inventory | undefined
+  >(undefined);
+
+  constructor(private localService: LocalNodeService, private pollService: PollService) {
+    this.localService
+      .status()
+      .pipe(this.pollService.poll((status: NodeStatus) => !status.inited))
+      .subscribe({
+        next: (status: NodeStatus) => {
+          console.log('checking node status: ', status);
+          if (status.inited) {
+            this.obtainInventory();
+          }
+        }
+      });
+
+    this.inventory$ = this.inventorySource.asObservable();
+  }
+
+  getDevices(): Observable<Volume[]> {
```

Comment:
I'm currently not sure how to progress here. If we stay on using this implementation we should add a documentation for this method which tells the developers that they subscribe to a subject. This is important because the subscription will be called automatically whenever the subject has been changed. This behavior can cause unexpected side-effects if this is not documented.

Otherwise we should use `ReplaySubject` which lead to the following implementation:
```
diff --git a/src/glass/src/app/shared/services/api/local-inventory.service.ts b/src/glass/src/app/shared/services/api/local-inventory.service.ts
--- a/src/glass/src/app/shared/services/api/local-inventory.service.ts	(revision a14335c9f96c8053c4bd43f9ee3ebd4320028ebc)
+++ b/src/glass/src/app/shared/services/api/local-inventory.service.ts	(date 1621954017930)
@@ -13,7 +13,7 @@
  * GNU General Public License for more details.
  */
 import { Injectable } from '@angular/core';
-import { BehaviorSubject, Observable } from 'rxjs';
+import { Observable, ReplaySubject } from 'rxjs';
 import { map } from 'rxjs/operators';
 
 import {
@@ -34,11 +34,9 @@
   providedIn: 'root'
 })
 export class LocalInventoryService {
-  public inventory$: Observable<Inventory | undefined>;
+  public inventory$: Observable<Inventory>;
 
-  private inventorySource: BehaviorSubject<Inventory | undefined> = new BehaviorSubject<
-    Inventory | undefined
-  >(undefined);
+  private inventorySource: ReplaySubject<Inventory> = new ReplaySubject<Inventory>(1);
 
   constructor(private localService: LocalNodeService, private pollService: PollService) {
     this.localService
@@ -57,22 +55,14 @@
   }
 
   getDevices(): Observable<Volume[]> {
-    return this.inventorySource.pipe(
-      map((inventory?: Inventory) => {
-        if (!inventory) {
-          return [];
-        }
-        return inventory.disks;
-      })
+    return this.inventory$.pipe(
+      map((inventory: Inventory) => inventory.disks)
     );
   }
 
   getNICs(): Observable<NICEntry[]> {
     return this.inventory$.pipe(
-      map((inventory?: Inventory) => {
-        if (!inventory) {
-          return [];
-        }
+      map((inventory: Inventory) => {
         const entries: NICEntry[] = [];
         Object.keys(inventory.nics).forEach((key: string) => {
           entries.push({
```

I think using `ReplaySubject` is the better approach because it behaves that way i expect when i subscribe to a method like `getDevices` and `getNICS`; they respond when the information/data is available.

Reply:
Some more thoughts. Is it intended to fetch the disks and nics only once and return that cached data on every call of `getDevices` and `getNICs`? If so, this is another + for using `ReplaySubject`.